doc: Add "debugging" section to the nRF5340 User Guide. (#6666)
nrfconnect/sdk-nrf: doc/nrf/ug_nrf5340.rst

Debugging
*********

If your want to debug just the application core firmware, you need a single debug session.
You can set it up as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.

However if you want to debug the firmware running on the network core you need to set up two separate debug sessions:

* A debug session with the application core.
* A debug session with a network core.

You can use the links above to set up each session.
Make sure to select the appropriate CPU for debugging: the nRF5340 application core or the nRF5340 network core.
When both sessions are established, first run the code execution on the application core.
Its startup code will release the "NETWORK.FORCEOFF" signal to start the network core and it will allocate the necessary GPIO pins for the network core.
Then you can start code execution on the network core in the other debug session.

.. note::
  When debugging if you want to reset the network core remember to reset and run code execution on the application core first.

.. note::
  You can also use alternative tools for debugging:
 
  * SEGGER Ozone.
  * GDB command line tool. 

> .. note::

Author: This is to be confirmed.

* A debug session with the application core.
* A debug session with the network core.

You can use the links above to set up each session.
Make sure to select the appropriate CPU for debugging: the nRF5340 application core or the nRF5340 network core.
When both sessions are established, first run the code execution on the application core.
Its startup code will release the "NETWORK.FORCEOFF" signal to start the network core and it will allocate the necessary GPIO pins for the network core.
Then you can start code execution on the network core in the other debug session.

.. figure:: /images/debugging_ozone_nrf5340_app.png
     :alt: SEGGER Ozone debugging session for the nRF5340 application core

     SEGGER Ozone debugging session for the nRF5340 application core

.. figure:: /images/debugging_ozone_nrf5340_net.png
     :alt: SEGGER Ozone debugging session for the nRF5340 network core

     SEGGER Ozone debugging session for the nRF5340 network core

.. note::
  When debugging if you want to reset the network core remember to reset and run code execution on the application core first.

.. note::
  You can also use alternative tools for debugging:
 

>   * SEGGER Ozone.

Reviewer: Should it be `Visual Studio Code` since you described debugging with SEGGER Ozone above ?

Author: Good point. I think this is already resolved through the doc review from @divipillai. The thing is, we don't want to describe the very details of setting up the 2 debug sessions with the recommended toolchains as this is either hard (VS Code, and they plan to automate it) or we are not sure it's possible (SES). So we mention those tools as an alternative.

The build system will automatically place both the application core and the network core update binaries (:file:`app_update.bin` and :file:`net_core_app_update.bin`) into a container package named :file:`dfu_application.zip`.
This container package can be used by update tools to pass both images during the simultaneous update of multiple images.

Debugging
*********

To debug the application core firmware, you need a single debug session.

> The debug session is set up as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.

Reviewer 2: ```suggestion
Set up the debug session as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```

> However, to debug the firmware running on the network core, you need to set up separate debug sessions for each core:
> 
> * Application core.
> * Network core.

Reviewer 2: ```suggestion
To debug the firmware running on the network core, you also need to set up separate debug sessions for the application core.
```

> Perform the following steps to start a debugging session in network core:

Reviewer 2: ```suggestion
Complete the following steps to start a debugging session in network core:
```

> 1. Set up each session for application and network core as mentioned in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.

Reviewer 2: ```suggestion
1. Set up a session for the application core and network core as mentioned in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```

#. Select the appropriate CPU for debugging, either the nRF5340 application core or the nRF5340 network core.

> #. Once both sessions are established, run the code execution on the application core.
>    The startup code will release the ``NETWORK.FORCEOFF`` signal to start the network core and it will allocate the necessary GPIO pins for the network core.

Reviewer 2: ```suggestion
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
```

#. Start code execution on the network core in the other debug session.

> 

Reviewer 2: Three consecutive notes is not ok. I believe one note only should be enough. Maybe the last one.

.. note::

>   While debugging, if you want to reset the network core, ensure to first reset and run code execution on the application core.

Reviewer 2: ```suggestion
  If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.
```

.. note::

>   You can use alternative tools for debugging:

Reviewer 2: ```suggestion
  You can use, for example, the following tools for debugging:
```
